problem_builder/public/js/mentoring_with_steps.js

```diff
@@ -0,0 +1,157 @@
+function MentoringWithStepsBlock(runtime, element) {
+
+    var steps = runtime.children(element).filter(
+        function(c) { return c.element.className.indexOf('pb-mentoring-step') > -1; }
+    );
+    var step = $('.mentoring', element).data('step');
+    var active_child, checkmark, submitDOM, nextDOM, tryAgainDOM, submitXHR;
+
+    function isLastChild() {
+        return (active_child === steps.length-1);
+    }
+
+    function updateStep() {
+        var handlerUrl = runtime.handlerUrl(element, 'update_step');
+        $.post(handlerUrl, JSON.stringify(step+1))
+            .success(function(response) {
+                step = response.step;
+            });
+    }
+
+    function handleResults(response) {
+        // Update step so next step is shown on page reload (even if user does not click "Next Step")
+        updateStep();
+
+        // Update UI
+        if (response.completed === 'correct') {
+            checkmark.addClass('checkmark-correct icon-ok fa-check');
+        } else if (response.completed === 'partial') {
+            checkmark.addClass('checkmark-partially-correct icon-ok fa-check');
+        } else {
+            checkmark.addClass('checkmark-incorrect icon-exclamation fa-exclamation');
+        }
+
+        submitDOM.attr('disabled', 'disabled');
+
+        nextDOM.removeAttr("disabled");
+        if (nextDOM.is(':visible')) { nextDOM.focus(); }
+
+        if (isLastChild()) {
+            tryAgainDOM.removeAttr('disabled');
+            tryAgainDOM.show();
+        }
+    }
+
+    function submit() {
+        // We do not handle submissions at this level, so just forward to "submit" method of current step
+        var child = steps[active_child];
+        child['submit'](handleResults);
+    }
+
+    function hideAllSteps() {
+        for (var i=0; i < steps.length; i++) {
+            $(steps[i].element).hide();
+        }
+    }
+
+    function cleanAll() {
+        checkmark.removeClass('checkmark-correct icon-ok fa-check');
+        checkmark.removeClass('checkmark-partially-correct icon-ok fa-check');
+        checkmark.removeClass('checkmark-incorrect icon-exclamation fa-exclamation');
+        hideAllSteps();
+    }
+
+    function displayNextChild() {
+        cleanAll();
+        findNextChild();
+        nextDOM.attr('disabled', 'disabled');
+        validateXBlock();
+    }
+
+    function findNextChild() {
+        // find the next real child block to display. HTMLBlock are always displayed
+        ++active_child;
+        var child = steps[active_child];
+        $(child.element).show();
```

Comment:
@itsjeyd nit: I would expect that `findNextChild` would return next child, not show it right away. Consider either renaming or returning the child and delegating showing it to the caller.